src/internal_modules/roc_core/mov_histogram.h

```diff
@@ -0,0 +1,111 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+//! @file roc_core/mov_histogram.h
+//! @brief Rolling window moving histogram.
+
+#ifndef ROC_CORE_MOV_HISTOGRAM_H_
+#define ROC_CORE_MOV_HISTOGRAM_H_
+
+#include "roc_core/array.h"
+#include "roc_core/iarena.h"
+#include "roc_core/panic.h"
+#include "roc_core/ring_queue.h"
+
+namespace roc {
+namespace core {
+
+/// @brief
+//! Efficiently implements moving histogram based on approach
+/// @tparam T The type of values to be stored in the histogram
+template <typename T> class MovHistogram {
+public:
+    //! Initialize.
+    MovHistogram(IArena& arena,
+                 T value_range_min,
+                 T value_range_max,
+                 size_t num_bins,
+                 size_t window_length)
+        : value_range_min_(value_range_min)
+        , value_range_max_(value_range_max)
+        , num_bins_(num_bins)
+        , window_length_(window_length)
+        , ring_buffer_(arena, window_length)
+        , bins_(arena)
+        , valid_(false) {
+        if (num_bins == 0 || window_length == 0 || value_range_min >= value_range_max) {
+            roc_panic("mov histogram: number of bins and window length must be greater "
+                      "than 0 and value_range_min must be less than value_range_max");
+        }
+
+        bin_width_ = (static_cast<T>(value_range_max - value_range_min)
+                      / static_cast<T>(num_bins));
+
+        if (!ring_buffer_.is_valid() || !bins_.resize(num_bins)) {
+            return;
+        }
+
+        valid_ = true;
+    }
+
+    //! Check if the histogram is valid.
+    bool is_valid() const {
+        return valid_;
+    }
+
+    //! Add a value to the histogram.
+    void add_value(const T& value) {
+        T clamped_value = value;
+
+        if (clamped_value < value_range_min_) {
+            clamped_value = value_range_min_;
+        } else if (clamped_value > value_range_max_) {
+            clamped_value = value_range_max_;
+        }
+
+        if (ring_buffer_.size() == window_length_) {
+            T oldest_value = ring_buffer_.front();
+            ring_buffer_.pop_front();
+            size_t oldest_bin_index = get_bin_index_(oldest_value);
+            if (oldest_bin_index < num_bins_) {
+                bins_[oldest_bin_index]--;
+            }
+        }
+
+        ring_buffer_.push_back(clamped_value);
+        size_t new_bin_index = get_bin_index_(clamped_value);
+        if (new_bin_index < num_bins_) {
+            bins_[new_bin_index]++;
+        }
+    }
+
+    //! returns the number of values in the bin
+    size_t get_bin_counter(size_t bin_index) const {
+        return bins_[bin_index];
+    }
+
+private:
+    //! returns the bin index for a given value
+    size_t get_bin_index_(const T& value) const {
+        return static_cast<size_t>((value - value_range_min_) / bin_width_);
```

Comment:
There is a corner case (see the right parenthesis my previous comment which mean a non-inclusive range: https://github.com/roc-streaming/roc-toolkit/pull/760#discussion_r1684971776): if `value == value_range_max`, bin index will be `num_bins_` (which means N+1), which is incorrect.

There are multiple solutions how to deal with it. One is to add a branch:
```
if (value == value_range_max_) {
    return num_bins_ - 1;
}
return ...
```

This should fix the problem in line 75. Also it would be nice to add a test for this case.